### PR TITLE
Added configArtifactRepository method into `sdk/python/kfp/compiler/compiler.py`

### PR DESCRIPTION
**Pull request by DmitryBe**
Added configArtifactRepository method into `sdk/python/kfp/compiler/compiler.py`

Example:
```
# config compiler (use s3)
cpl = compiler.Compiler() \
    .configArtifactRepository(
        endpoint='s3.amazonaws.com',
        bucket='<bucket>', 
        keyPrefix='pipelines/dev',
        accessKey='', secretKey=''
    )

pipeline_filename = pipeline_func.__name__ + '.pipeline.tar.gz'
cpl.compile(pipeline_func, pipeline_filename)

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/765)

### CODE REVIEW

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -35,8 +35,32 @@ def my_pipeline(a: dsl.PipelineParam, b: dsl.PipelineParam):
     pass
 
   Compiler().compile(my_pipeline, 'path/to/workflow.yaml')
+
+  # config artifact repo to use S3
+  .configArtifactRepository(
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py]: Are you sure this code works?

File: sdk/python/Makefile [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
+build:
+	./build.sh
+
+upload-to-s3:
```
Comment (Ark-kun) [inline sdk/python/Makefile]: This file probably should not be a part of this PR.

File: sdk/python/kfp/_client.py [hunk 3]
```diff
@@ -28,11 +28,8 @@
 class Client(object):
```
Comment (Ark-kun) [inline sdk/python/kfp/_client.py]: The changes to this class probably belong to a different PR.

**Comment by DmitryBe** [comment]
Hi @Ark-kun, 

Thank you for your review. To make it clear, in order to add AWS S3 support for storing artifacts, I have made 3 PR in total:

1. backend extend (https://github.com/kubeflow/pipelines/pull/766)
2. frontend extend (https://github.com/kubeflow/pipelines/pull/767)
3. the current one, python DSL extend.

To extend python DSL, i have added `configArtifactRepository` method, to parametrize method `_build_conventional_artifact` (default minio values where hardcoded previously).

# example: 
Compiler() \
  .configArtifactRepository(
      endpoint='s3.amazonaws.com',
      bucket='<your-bucket-name>', 
      keyPrefix='runs/dev',
      accessKey='', secretKey=''
  ) \
.compile(my_pipeline, 'path/to/workflow.yaml')

I have tested this code.

Dmitry,

File: sdk/python/Makefile (hunk 2 above)
Comment (DmitryBe) [reply sdk/python/Makefile]: removed

File: sdk/python/kfp/_client.py (hunk 3 above)
Comment (DmitryBe) [reply sdk/python/kfp/_client.py]: excluded

**Comment by DmitryBe** [comment]
Hi @Ark-kun,
I cleaned PR based on your suggestions, left only compiler related changes (method `configArtifactRepository` to parametrize `_build_conventional_artifact`)

File: sdk/python/kfp/compiler/compiler.py [hunk 4]
```diff
@@ -557,6 +584,19 @@ def _compile(self, pipeline_func):
     workflow = self._create_pipeline_workflow(args_list_with_defaults, p)
     return workflow
 
+  def configArtifactRepository(self, endpoint='minio-service.kubeflow:9000', bucket='mlpipeline', keyPrefix='runs', accessKey='accesskey', secretKey='secretkey', insecure=True):
+    """Config artifact repository
+    """
+    self.artifactRepositoryConfig = Compiler.ArtifactRepositoryConfig(
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py]: Can you make this field private (leading underscore)?

File: sdk/python/kfp/compiler/compiler.py [hunk 5]
```diff
@@ -557,6 +584,19 @@ def _compile(self, pipeline_func):
     workflow = self._create_pipeline_workflow(args_list_with_defaults, p)
     return workflow
 
+  def configArtifactRepository(self, endpoint='minio-service.kubeflow:9000', bucket='mlpipeline', keyPrefix='runs', accessKey='accesskey', secretKey='secretkey', insecure=True):
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py]: Python uses `snake_case` for function names. `def set_artifact_repository(...)`

**Comment by Ark-kun** [comment]
@IronPan @DmitryBe Do you think that it would be better to configure this at the cluster level and not bake this information into the pipeline file?

I think a better solution would be for the backend to read the default artifact repository information from cluster's ConfigMap and add it to the workflow at submission time. WDYT?

**Comment by vicaire** [comment]
@DmitryBe, @Ark-kun, what is the status on this PR? Is it ready for another round of review? Thanks.

**Comment by Ark-kun** [comment]
BTW, I had a PR where I've made the artifact storage configurable on the cluster level and made the compiled pipelines portable: https://github.com/kubeflow/pipelines/pull/513

**Comment by Ark-kun** [comment]
The support for setting the artifact storage options was merged in different PRs.

**Comment by Ark-kun** [comment]
This functionality has already been merged.
